File: answers/Remove Islands/Python/removeIslands.py
def removeIslands(matrix):
    # Write your code here.

    # 1. Iterate over each value that is on the border
    numRows = len(matrix)
    numCols = len(matrix[0])

    for row in range(numRows):
        for col in range(numCols):
            border = row==0 or col==0 or row==numRows-1 or col==numCols-1
            if border:
                if matrix[row][col] == 0:
                    print(f"there is nothing to do with 0's")
                if matrix[row][col] == 1:
                    print(f"we found a 1 on the border, need to use the stack to block off its neighbors")

                    # 2. take the values of 1 and look at its neighbors, turn everything into 2's
                    turn1to2(matrix, row, col)
                if matrix[row][col] == 2:
                    print(f"we already took care of this, that is why it is a 2")

    # 3. Finish by turning 2's back to 1's, 1's back to 0's
    for row in range(numRows):
        for col in range(numCols):  
            val = matrix[row][col]
            if val==1:
                matrix[row][col] = 0
            if val==2:
                matrix[row][col] = 1

    print(f"matrix: {matrix}")
    return matrix

def turn1to2(matrix, row, col):
    stack = []
    stack.append([row, col])

    while len(stack) > 0:
        currentPosition = stack.pop()
        currentRow = currentPosition[0]
        currentCol = currentPosition[1]

        # change current value to 2 
        matrix[currentRow][currentCol] = 2

        # look for neighbors that should be added to the stack
        # Make sure of the following:
        # 1. the neighboor is 'in bounds ' to prevent indexing errors
        # 2. the neighbor is a 1 (2's are already dealt with, 0's are irrelevant)
        neighbors = getNeighbors(matrix, currentRow, currentCol)
        for neighbor in neighbors:
            row = neighbor[0]
            col = neighbor[1]
            if matrix[row][col] == 1:
                stack.append(neighbor)
            
            
def getNeighbors(matrix, row, col):
    neighbors = []

    numRows = len(matrix)
    numCols = len(matrix[0])

    #up
    if row - 1 >= 0:
        neighbors.append([row - 1, col])

    #down
    if row + 1 < numRows:
        neighbors.append([row + 1, col])

    #left
    if col - 1 >= 0:
        neighbors.append([row, col - 1])

    #right
    if col + 1 < numCols:
        neighbors.append([row, col + 1])

    return neighbors

    
    
input = [
    [1, 0, 0, 0, 0, 0],
    [0, 1, 0, 1, 1, 1],
    [0, 0, 1, 0, 1, 0],
    [1, 1, 0, 0, 1, 0],
    [1, 0, 1, 1, 0, 0],
    [1, 0, 0, 0, 0, 1],
]
expected = [
    [1, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 1],
    [0, 0, 0, 0, 1, 0],
    [1, 1, 0, 0, 1, 0],
    [1, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 1],
]
actual = removeIslands(input)
assert(actual == expected)
print("\n\nAssertion passed, good work!\n\n")
    
                

File: AlgoExpert/coding_questions/Remove Islands/C++/removeIslands.cpp
#include <vector>
#include <cassert> // needed for assert()
#include <cstdio> // needed for printf()
using namespace std;

vector<vector<int>> getNeighbors(int row, int col, vector<vector<int>> matrix) {
  int numRows = matrix.size();
  int numCols = matrix[0].size();

  vector<vector<int>> neighbors;
  if (row - 1 >= 0) {neighbors.push_back(vector<int>{row-1, col});} //up
  if (row + 1 < numRows) {neighbors.push_back(vector<int>{row+1, col});} //down
  if (col - 1 >= 0) {neighbors.push_back(vector<int>{row, col-1});} //left
  if (row + 1 < numCols) {neighbors.push_back(vector<int>{row, col+1});} //right

  //don't forget to return lol
  return neighbors;
}

void turn1to2(int row, int col, vector<vector<int>> &matrix) {// This function EDITS matrix, make sure input takes a reference ie. '&matrix'
  vector<vector<int>> stack;
  stack.push_back(vector<int>{row, col});

  while (stack.size() > 0) {
    auto currentPosition = stack[stack.size() - 1];
    stack.pop_back();
    int currentRow = currentPosition[0];
    int currentCol = currentPosition[1];

    matrix[currentRow][currentCol] = 2;
    
    auto neighbors = getNeighbors(currentRow, currentCol, matrix);
    for (auto neighbor: neighbors) {
      int row = neighbor[0];
      int col = neighbor[1];
      if (matrix[row][col] == 1) {stack.push_back(neighbor);}
    }
    
  }
 
}

vector<vector<int>> removeIslands(vector<vector<int>> matrix) {
  // Write your code here.

  // 1. Iterate through each position that is on the border, looking for connected positions 
  int numRows = matrix.size();
  int numCols = matrix[0].size();
  for (int row = 0; row < numRows; row++) {
    for (int col = 0; col < numCols; col++) {
      if (row==0 || col==0 || row==numRows-1 || col==numCols-1) {
        if (matrix[row][col] == 1) {
          // 2. Turn 1's on the border, and anything touching them, into 2's
          turn1to2(row, col, matrix);
        }
      }
    }
  }

  // 3. Turn 2's (connected to 1's on the border) into 1's, turn 1's (islands that need to be removed) into 0's
  for (int row = 0; row < numRows; row++) {
    for (int col = 0; col < numCols; col++) {
      if (matrix[row][col] == 1) {matrix[row][col] = 0;}
      if (matrix[row][col] == 2) {matrix[row][col] = 1;}
    }
  }
  return matrix;
}


int main() {
    vector<vector<int>> input = {
        {1, 0, 0, 0, 0, 0}, 
        {0, 1, 0, 1, 1, 1}, 
        {0, 0, 1, 0, 1, 0},
        {1, 1, 0, 0, 1, 0}, 
        {1, 0, 1, 1, 0, 0}, 
        {1, 0, 0, 0, 0, 1},
    };
    vector<vector<int>> expected = {
        {1, 0, 0, 0, 0, 0}, 
        {0, 0, 0, 1, 1, 1}, 
        {0, 0, 0, 0, 1, 0},
        {1, 1, 0, 0, 1, 0}, 
        {1, 0, 0, 0, 0, 0}, 
        {1, 0, 0, 0, 0, 1},
    };

    auto actual = removeIslands(input);
    assert(expected == actual);
    printf("\n\nAssertion passed, good work!\n\n");

    return 0;
}


